x/staking/keeper/liquid_stake.go

```diff
@@ -0,0 +1,395 @@
+	timeKey := types.GetTokenizeShareAuthorizationTimeKey(completionTime)
+	bz := k.cdc.MustMarshal(&authorizations)
+	store.Set(timeKey, bz)
+}
+
+// Returns a list of addresses pending tokenize share unlocking at the same time
+func (k Keeper) GetPendingTokenizeShareAuthorizations(ctx sdk.Context, completionTime time.Time) types.PendingTokenizeShareAuthorizations {
+	store := ctx.KVStore(k.storeKey)
+
+	timeKey := types.GetTokenizeShareAuthorizationTimeKey(completionTime)
+	bz := store.Get(timeKey)
+
+	authorizations := types.PendingTokenizeShareAuthorizations{Addresses: []string{}}
+	if len(bz) == 0 {
+		return authorizations
+	}
+	k.cdc.MustUnmarshal(bz, &authorizations)
+
+	return authorizations
+}
+
+// Inserts the address into a queue where it will sit for 1 unbonding period
+// before the tokenize share lock is removed
+// Returns the completion time
+func (k Keeper) QueueTokenizeSharesAuthorization(ctx sdk.Context, address sdk.AccAddress) time.Time {
+	params := k.GetParams(ctx)
+	completionTime := ctx.BlockTime().Add(params.UnbondingTime)
+
+	// Append the address to the list of addresses that also unlock at this time
+	authorizations := k.GetPendingTokenizeShareAuthorizations(ctx, completionTime)
+	authorizations.Addresses = append(authorizations.Addresses, address.String())
+
+	k.SetPendingTokenizeShareAuthorizations(ctx, completionTime, authorizations)
+	k.SetTokenizeSharesUnlockTime(ctx, address, completionTime)
+
+	return completionTime
+}
+
+// Cancels a pending tokenize share authorization by removing the lock from the queue
+func (k Keeper) CancelTokenizeShareLockExpiration(ctx sdk.Context, address sdk.AccAddress, completionTime time.Time) {
+	authorizations := k.GetPendingTokenizeShareAuthorizations(ctx, completionTime)
+
+	updatedAddresses := []string{}
+	for _, expiringAddress := range authorizations.Addresses {
+		if address.String() != expiringAddress {
+			updatedAddresses = append(updatedAddresses, expiringAddress)
+		}
+	}
+
+	authorizations.Addresses = updatedAddresses
+	k.SetPendingTokenizeShareAuthorizations(ctx, completionTime, authorizations)
+}
+
+// Unlocks all queued tokenize share authorizations that have matured
+// (i.e. have waited the full unbonding period)
+func (k Keeper) RemoveExpiredTokenizeShareLocks(ctx sdk.Context, blockTime time.Time) (unlockedAddresses []string) {
+	store := ctx.KVStore(k.storeKey)
+
+	// iterators all time slices from time 0 until the current block time
+	prefixEnd := sdk.InclusiveEndBytes(types.GetTokenizeShareAuthorizationTimeKey(blockTime))
+	iterator := store.Iterator(types.TokenizeSharesUnlockQueuePrefix, prefixEnd)
+	defer iterator.Close()
+
+	unlockedAddresses = []string{}
+	for ; iterator.Valid(); iterator.Next() {
+		authorizations := types.PendingTokenizeShareAuthorizations{}
+		k.cdc.MustUnmarshal(iterator.Value(), &authorizations)
+
+		for _, addressString := range authorizations.Addresses {
+			k.RemoveTokenizeSharesLock(ctx, sdk.MustAccAddressFromBech32(addressString))
```

Comment:
Not sure it is safe to delete inside an iterator. cc @alexanderbez

Reply:
I assumed it was safe as validator unbonding's use the [same mechanism ](https://github.com/iqlusioninc/cosmos-sdk/blob/v0.45.16-ics-lsm/x/staking/keeper/validator.go#L399), and the unit test passed. But also curious to hear @alexanderbez 's opinion

Reply:
I was under the impression that you want to avoid writes to a keyspace during iteration, but given that we already do it in some places, perhaps I'm wrong.

Reply:
I'll just move the Removal to outside of the loop to be safe

Reply:
Resolved [here](https://github.com/cosmos/cosmos-sdk/pull/16747/commits/04843f4ab98af92937f585a39e99d28623dc68dc)